### PR TITLE
Port from the private repo

### PR DESCRIPTION
**Pull request by qinlz2**
- Port from the private repo
- Rebase to upstream `develop` branch

### CODE REVIEW

File: beacon-chain/rpc/apimiddleware/custom_handlers.go [hunk 1]
```diff
@@ -403,6 +404,15 @@ func receiveEvents(eventChan <-chan *sse.Event, w http.ResponseWriter, req *http
 				data = &EventChainReorgJson{}
 			case events.SyncCommitteeContributionTopic:
 				data = &SignedContributionAndProofJson{}
+			case events.LightClientFinalityUpdateTopic:
+				data = &LightClientFinalityUpdateResponseJson{}
+			case events.LightClientOptimisticUpdateTopic:
+				data = &LightClientOptimisticUpdateResponseJson{}
+			// Below is ChainSafe's format
```
Comment (7AC) [inline beacon-chain/rpc/apimiddleware/custom_handlers.go]: remove?

File: beacon-chain/rpc/apimiddleware/structs.go [hunk 2]
```diff
@@ -8,9 +8,9 @@ import (
 	ethpbv2 "github.com/prysmaticlabs/prysm/v3/proto/eth/v2"
 )
 
-//----------------
+// ----------------
```
Comment (7AC) [inline beacon-chain/rpc/apimiddleware/structs.go:11]: remove?

File: beacon-chain/rpc/apimiddleware/structs.go [hunk 3]
```diff
@@ -302,9 +337,9 @@ type ForkchoiceResponse struct {
 	ForkChoiceNodes               []*ForkChoiceNodeJson `json:"forkchoice_nodes"`
 }
 
-//----------------
+// ----------------
```
Comment (7AC) [inline beacon-chain/rpc/apimiddleware/structs.go:340]: remove space?

File: beacon-chain/rpc/eth/beacon/light_client.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	"github.com/prysmaticlabs/prysm/v3/encoding/bytesutil"
+	ethpbv1 "github.com/prysmaticlabs/prysm/v3/proto/eth/v1"
+	ethpbv2 "github.com/prysmaticlabs/prysm/v3/proto/eth/v2"
+	"github.com/prysmaticlabs/prysm/v3/proto/migration"
+	"github.com/prysmaticlabs/prysm/v3/time/slots"
+)
+
+// GetLightClientBootstrap -
```
Comment (7AC) [inline beacon-chain/rpc/eth/beacon/light_client.go]: comment referencing the spec?

File: beacon-chain/rpc/eth/beacon/light_client.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	}
+
+	// Prepare header
+	signedBeaconHeader, err := blk.Header()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get block header: %v", err)
+	}
+	header := migration.V1Alpha1SignedHeaderToV1(signedBeaconHeader).GetMessage()
+
+	// Alternative way to prepare header (by not using migration pkg)
```
Comment (7AC) [inline beacon-chain/rpc/eth/beacon/light_client.go]: delete?

File: beacon-chain/rpc/eth/beacon/light_client.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+	result := &ethpbv2.LightClientBootstrapResponse{
+		Version: ethpbv2.Version(blk.Version()),
+		Data:    &data,
+	}
+
+	return result, nil
+}
+
+// GetLightClientFinalityUpdate -
```
Comment (7AC) [inline beacon-chain/rpc/eth/beacon/light_client.go]: comment?

File: beacon-chain/rpc/eth/beacon/light_client.go [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+	// Return the result
+	result := &ethpbv2.LightClientFinalityUpdateResponse{
+		Version: ethpbv2.Version(headBlk.Version()),
+		Data:    &data,
+	}
+
+	return result, nil
+}
+
+// GetLightClientOptimisticUpdate -
```
Comment (7AC) [inline beacon-chain/rpc/eth/beacon/light_client.go]: comment?

File: beacon-chain/rpc/eth/events/events.go [hunk 8]
```diff
@@ -205,6 +213,43 @@ func handleStateEvents(
...
+	case statefeed.LightClientOptimisticUpdate:
+		if _, ok := requestedTopics[LightClientOptimisticUpdateTopic]; !ok {
+			return nil
+		}
+		update, ok := event.Data.(*ethpbv2.LightClientOptimisticUpdateResponse)
+		if !ok {
+			return nil
+		}
+		return streamData(stream, LightClientOptimisticUpdateTopic, update)
+	// Below is ChainSafe's format
```
Comment (7AC) [inline beacon-chain/rpc/eth/events/events.go]: delete?

File: proto/eth/service/beacon_chain_service.proto [hunk 9]
```diff
@@ -419,5 +420,32 @@ service BeaconChain {
...
+  // Spec: https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getLightClientBootstrap
+  rpc GetLightClientBootstrap(v2.LightClientBootstrapRequest) returns (v2.LightClientBootstrapResponse) {
+    option (google.api.http) = {
+      get: "/internal/eth/v1/beacon/light_client/bootstrap/{block_root}"
+    };
+  }
+
+  // GetLightClientFinalityUpdate retrieves the latest finality update for light client
+  //
+  // Spec: https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getLightClientFinalityUpdate
```
Comment (7AC) [inline proto/eth/service/beacon_chain_service.proto:435]: maybe reference a version here?

File: proto/eth/service/beacon_chain_service.proto [hunk 10]
```diff
@@ -419,5 +420,32 @@ service BeaconChain {
...
+  rpc GetLightClientFinalityUpdate(google.protobuf.Empty) returns (v2.LightClientFinalityUpdateResponse) {
+    option (google.api.http) = {
+      get: "/internal/eth/v1/beacon/light_client/finality_update"
+    };
+  }
+
+  // GetLightClientOptimisticUpdate retrieves the latest finality update for light client
+  //
+  // Spec: https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getLightClientOptimisticUpdate
```
Comment (7AC) [inline proto/eth/service/beacon_chain_service.proto:444]: ditto

File: proto/ssz_proto_library.bzl [hunk 11]
```diff
@@ -15,11 +15,14 @@ mainnet = {
     "previous_epoch_attestations.max": "4096",  # MAX_ATTESTATIONS * SLOTS_PER_EPOCH
     "current_epoch_attestations.max": "4096",  # MAX_ATTESTATIONS * SLOTS_PER_EPOCH
     "slashings.size": "8192",  # EPOCHS_PER_SLASHINGS_VECTOR
-    "sync_committee_bits.size": "512", #SYNC_COMMITTEE_SIZE
+    "sync_committee_bits.size": "512",  #SYNC_COMMITTEE_SIZE
```
Comment (7AC) [inline proto/ssz_proto_library.bzl]: remove space

File: beacon-chain/rpc/apimiddleware/custom_handlers.go (hunk 1 above)
Comment (qinlz2) [reply beacon-chain/rpc/apimiddleware/custom_handlers.go]: Done

File: beacon-chain/rpc/apimiddleware/structs.go (hunk 2 above)
Comment (qinlz2) [reply beacon-chain/rpc/apimiddleware/structs.go:11]: This is from the original code.   (Only a space is added by IDE, so let's don't do it)

File: beacon-chain/rpc/apimiddleware/structs.go (hunk 3 above)
Comment (qinlz2) [reply beacon-chain/rpc/apimiddleware/structs.go:340]: Ditto.

File: beacon-chain/rpc/eth/beacon/light_client.go (hunk 4 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/beacon/light_client.go]: Done

File: beacon-chain/rpc/eth/beacon/light_client.go (hunk 5 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/beacon/light_client.go]: Done

File: beacon-chain/rpc/eth/events/events.go (hunk 8 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/events/events.go]: Done

File: beacon-chain/rpc/eth/beacon/light_client.go (hunk 7 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/beacon/light_client.go]: Done

File: proto/ssz_proto_library.bzl (hunk 11 above)
Comment (qinlz2) [reply proto/ssz_proto_library.bzl]: IDE auto did this.   (Let's keep it, otherwise, it will do the same again when modify another time)

Comment (7AC) [reply proto/ssz_proto_library.bzl]: Can you change the IDE config? I think it's better to minimize changes

File: beacon-chain/rpc/eth/beacon/light_client.go (hunk 6 above)
Comment (qinlz2) [reply beacon-chain/rpc/eth/beacon/light_client.go]: Done

File: proto/ssz_proto_library.bzl (hunk 11 above)
Comment (qinlz2) [reply proto/ssz_proto_library.bzl]: Yes, I can.    

But the commenting style with and without initial space exists in `prysm` codebase. 
IDE does not support per-file settings.    

If change IDE, some other files will be "auto" impacted as well....    One sample is like the one below from the original code.  
`// SignedBeaconBlock is an interface describing the method set of
// a signed beacon block.
type SignedBeaconBlock interface {
	Block() BeaconBlock
	Signature() [field_params.BLSSignatureLength]byte
	IsNil() bool
`

Comment (qinlz2) [reply proto/ssz_proto_library.bzl]: Yes, for this file (.bzl) I removed the extra white space to match the original file. 

   - Within this file, looks like there are inconsistent comment style (with / without leading space already).  
   
   - For code review, I found an option to "Hide whitespace", in the setting icon near the top left.

Comment (7AC) [reply proto/ssz_proto_library.bzl]: I would just avoid the change if possible, think also about conflicts etc. Maybe disable the feature in your editor for the time being?

**Review by 7AC — APPROVED** [review]

**Pull request merged**